conwayste/src/ui/treeview.rs

```diff
@@ -0,0 +1,309 @@
+    }
+
+    /// Get a mutable reference to a Node. The specified Node must be accessible.
+    pub fn get_mut(&mut self, node_id: &NodeId) -> Result<&mut Node<T>, Box<dyn Error>> {
+        if !self.can_access(node_id)? {
+            return Err("the TreeView does not have access to the specified Node".into());
+        }
+
+        let tree = unsafe { self.tree.as_mut() };
+
+        tree.get_mut(node_id).map_err(|e| e.into())
+    }
+
+    /// Inserts a node into the underlying Tree. See documentation for Tree.insert.
+    ///
+    /// # Errors
+    ///
+    /// In addition to errors from Tree.insert, an error is returned when the TreeView is
+    /// sharing the Tree.
+    #[allow(unused)]
+    pub fn insert(
+        &mut self,
+        node: Node<T>,
+        behavior: InsertBehavior,
+    ) -> Result<NodeId, Box<dyn Error>> {
+        if !self.restriction.is_none() {
+            return Err("the TreeView must have full access to the Tree to insert a node".into());
+        }
+
+        let tree = unsafe { self.tree.as_mut() };
+
+        Ok(tree.insert(node, behavior)?)
+    }
+
+    /// Removes a node from the underlying Tree. See documentation for Tree.remove_node.
+    ///
+    /// # Errors
+    ///
+    /// In addition to errors from Tree.remove_node, an error is returned when the TreeView is
+    /// sharing the Tree.
+    #[allow(unused)]
+    pub fn remove(
+        &mut self,
+        node_id: NodeId,
+        behavior: RemoveBehavior,
+    ) -> Result<Node<T>, Box<dyn Error>> {
+        if !self.restriction.is_none() {
+            return Err("the TreeView must have full access to the Tree to remove a node".into());
+        }
+
+        let tree = unsafe { self.tree.as_mut() };
+
+        Ok(tree.remove_node(node_id, behavior)?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use id_tree::InsertBehavior;
+
+    #[test]
+    fn test_subtree_and_children_ids() {
+        let mut tree: Tree<i64> = id_tree::TreeBuilder::new().build();
+        let node_r = Node::new(0);
+        let root_node_id = tree.insert(node_r, InsertBehavior::AsRoot).unwrap();
+
+        let child_nodes = vec![Node::new(1), Node::new(2), Node::new(3)];
+
+        for c in child_nodes.into_iter() {
+            tree.insert(c, InsertBehavior::UnderNode(&root_node_id))
+                .unwrap();
+        }
+
+        let mut view = TreeView::new(&mut tree);
+
+        let (root_node_ref, mut sub_tree_ref) = view.sub_tree(&root_node_id).unwrap();
+        for child_id in sub_tree_ref.children_ids().iter() {
+            let child_ref = sub_tree_ref.get_mut(child_id).unwrap().data_mut();
+            // root += child
+            *root_node_ref.data_mut() += *child_ref;
+            // zero out child
+            *child_ref = 0;
+        }
+        drop(view);
+        assert_eq!(*tree.get(&root_node_id).unwrap().data(), 1 + 2 + 3);
+    }
+
+    /*
+    #[test]
+    fn test_compile_fail_double_mut_borrow() {
```

Comment:
Will this test ever be needed?

Reply:
Uncommenting it causes a compile failure -- if it doesn't, something got screwed up with borrowing.

I don't know of a way to have a test that verifies it can't compile, so I'm just leaving it commented.

Reply:
BTW if you find a way to have a test with code that only passes if it fails to compile, that would be amazing! :grin:

Reply:
This looks promising: https://crates.io/crates/compiletest_rs
but I think it's overkill for this one test. Fine with leaving it as is.